consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/MessageTracker.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft;
+
+import static java.util.Collections.newSetFromMap;
+
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.p2p.api.MessageData;
+
+import java.util.Set;
+
+public class MessageTracker {
+  private final Set<Hash> seenMessages;
+
+  public MessageTracker(final int gossipHistoryLimit) {
+    this.seenMessages = newSetFromMap(new SizeLimitedMap<>(gossipHistoryLimit));
```

Comment:
`WeakHashMap` will drop messages surprisingly often even if the JVM hasn't hit its full memory allocation yet.  Anytime it might have to expand the memory currently in use it will drop weak references first, so you can never count on anything being kept. So it's only ever suitable for a cache where values can be recreated on demand and even then I'd avoid it if possible because it makes GC more expensive.

Reply:
As AJS said, I want a stronger guarantee. And to ensure memory usage doesn't explode the size of the Map is limited.